KAB-1057: Add flows to MCP Server

radektomasek opened the pull request:
## Add Flow Management Tools to Keboola MCP Server

### Summary

Adds flow (orchestration) management functionality to enable creating and managing data pipelines through the MCP server.

### New Tools Added

- `create_flow` - Create new data flows/pipelines.
- `retrieve_flows` - List all flows or get specific flows by ID.
- `update_flow` - Modify existing flow configurations.
- `get_flow_detail` - Get detailed flow information.

### Files Added

- `keboola_mcp_server/tools/flows.py`

### Key Features

- Schema-compliant with official Keboola flow structure.
- Auto-generates unique IDs for phases and tasks.
- Validates dependencies and prevents circular references.
- Should work with any component type (extractors, transformations, writers) - Would like to do more testing on that front.

### Testing

- Successfully created, retrieved, and updated flows with AWS S3 extractor component

### Video

https://github.com/user-attachments/assets/8ea2aad9-6609-40c2-b66a-10db841c7546

radektomasek commented on src/keboola_mcp_server/tools/flow.py:
```diff
@@ -0,0 +1,365 @@
+                'parameters': flow_parameters
+            }
+        }
+    )
```
👍

radektomasek commented on src/keboola_mcp_server/tools/flow.py:
```diff
@@ -0,0 +1,365 @@
+                LOG.warning(f'Could not retrieve flow {flow_id}: {e}')
+        return flows
+    else:
+        raw_flows = await client.storage_client.configuration_list(component_id=ORCHESTRATOR_COMPONENT_ID)
```
👍

radektomasek commented on src/keboola_mcp_server/tools/flow.py:
```diff
@@ -0,0 +1,365 @@
+    phases: Annotated[List[Dict[str, Any]], Field(description='Updated list of phase definitions')],
+    tasks: Annotated[List[Dict[str, Any]], Field(description='Updated list of task definitions')],
+    change_description: Annotated[str, Field(description='Description of changes made')],
+) -> Annotated[ComponentRootConfiguration, Field(description='Updated flow configuration')]:
```
👍

radektomasek commented on src/keboola_mcp_server/tools/flow.py:
```diff
@@ -0,0 +1,365 @@
+    """
+    client = KeboolaClient.from_state(ctx.session.state)
+
+    raw_config = await client.storage_client.get(
```
👍

radektomasek commented on src/keboola_mcp_server/tools/flow.py:
```diff
@@ -0,0 +1,365 @@
+    """Ensure all tasks have unique IDs and proper structure according to schema"""
+    processed_tasks = []
+    used_ids = set()
+    task_counter = 20001
```
This pattern comes from the Kai-Bot implementation that was shared by @davidesner in the KAB-1057: https://github.com/keboola/kai-bot/blob/main/src/keboola/kaibot/backend/flow_backend.py

I just iterated with AI and tried to get some common ground for our implementation (including the incorporation of the Flow model).

The thoughts about this specific code:

```
The Logic:

Avoid conflicts with phase IDs - Phases typically use small numbers (1, 2, 3...).
Clear separation - Tasks get a completely different ID range.
Room for growth - 20,000+ IDs available for tasks.
```

The 20001 starting point is essentially a namespace separation technique - it's a simple, effective way to ensure phase and task IDs never collide while maintaining human-readable sequential numbering.

But in the first place - it was my starting point from the kaibot code. 

I will add some summary to the code itself. @mariankrotil

ncapek commented on src/keboola_mcp_server/tools/components/model.py:
```diff
@@ -311,3 +313,141 @@ class ComponentWithConfigurations(BaseModel):
+        )
+
+
+class FlowSummary(BaseModel):
```
We've been using the naming "Reduced" for the listing versions, e.g. "ReducedComponent", "ReducedConfiguration". If possible, can we be consistent with this?

radektomasek commented on src/keboola_mcp_server/tools/components/model.py:
```diff
@@ -311,3 +313,141 @@ class ComponentWithConfigurations(BaseModel):
+    tasks: List[FlowTask] = Field(description='List of tasks in the flow')
+
+
+class FlowConfigurationResponse(ComponentConfigurationResponseBase):
```
Great point about the response objects. You're absolutely right that mixing core domain objects with API response metadata creates unnecessary complexity.

I initially followed the existing component tools pattern (with the help of AI) for consistency, but I agree this is a good opportunity to improve the architecture.

I am happy to adjust the code and have a cleaner apprach while returning core objects only:

```python
# Instead of FlowConfigurationResponse
def create_flow(...) -> FlowConfiguration
def update_flow(...) -> FlowConfiguration  
def get_flow_detail(...) -> FlowConfiguration
def retrieve_flows(...) -> List[FlowSummary]
```

Let me address it now.

radektomasek commented:
@ncapek @mariankrotil I've updated the branch and addressed your latest PR comments, including the schema adjustment that @davidesner [mentioned in the ticket](https://keboola.atlassian.net/browse/KAB-1057?focusedCommentId=163978).

## Changes Made:
- ✅ **Fixed configuration structure** - Removed extra `parameters` wrapper (flows store phases/tasks directly under `configuration`).
- ✅ **Added comprehensive schema documentation** - Embedded the simplified flow schema in tool descriptions so LLMs have clear guidance.
- ✅ **Updated model parsing** - Fixed parsing to read from correct API response location.
- ✅ **Clean architecture maintained** - Tools return core objects only (`FlowConfiguration`, `List[ReducedFlow]`).

## Testing:
- **Local testing successful** - Flows create, retrieve, and update correctly using Claude as the MCP client.
- **Unit tests added** - Comprehensive test coverage including edge cases, validation logic, and integration scenarios.
- **All major functionality verified** - ID generation, circular dependency detection, and API integration working as expected.

The implementation now works "like a charm" as David mentioned - the LLM receives proper schema guidance and creates the correct API structure without guessing.

Ready for handover if you're satisfied with the changes! 🚀

cc @jordanrburger

mariankrotil commented on src/keboola_mcp_server/tools/flow.py:
```diff
@@ -0,0 +1,357 @@
+from keboola_mcp_server.client import KeboolaClient
+from keboola_mcp_server.errors import tool_errors
+from keboola_mcp_server.mcp import with_session_state
+from keboola_mcp_server.tools.components.model import (
```
As the files are getting large, We could create a new tools submodule: `tools/flows` where we could have `flows/model.py` - flow models, `flows/utils.py` - validation methods and `flows/tools.py` - tool implementations like it is done with components?

vita-stejskal replied:
Let's leave this refactoring to another PR.

mariankrotil commented on src/keboola_mcp_server/tools/flow.py:
```diff
@@ -0,0 +1,357 @@
+    description: Annotated[str, Field(description='Detailed description of the flow purpose')],
+    phases: Annotated[List[Dict[str, Any]], Field(description=f"""List of phase definitions.
+
+{FLOW_SCHEMA}
```
We should not add schemas to the descriptions as when the agent wants to list the tools, this tool description gets extremely large and can shadow others. @ncapek suggested tool for getting flow schmeas

vita-stejskal replied:
This is addressed by #124 .

mariankrotil commented on src/keboola_mcp_server/tools/flow.py:
```diff
@@ -0,0 +1,357 @@
+    configuration_id: Annotated[str, Field(description='ID of the flow configuration to update')],
+    name: Annotated[str, Field(description='Updated flow name')],
+    description: Annotated[str, Field(description='Updated flow description')],
+    phases: Annotated[List[Dict[str, Any]], Field(
+        description=f"""Updated list of phase definitions.
+            {FLOW_SCHEMA}
+            """
+        )],
+        tasks: Annotated[List[Dict[str, Any]], Field(
+            description=f"""Updated list of task definitions.
+                {FLOW_SCHEMA}"""
+        )],
+        change_description: Annotated[str, Field(description='Description of changes made')],
```
align the indentation, please

mariankrotil commented on src/keboola_mcp_server/client.py:
```diff
@@ -524,6 +526,85 @@ async def update_component_row_configuration(
+        """
+        return await self.configuration_list(component_id=ORCHESTRATOR_COMPONENT_ID)
+
+    async def get_flow_configuration(self, config_id: str) -> JsonDict:
```
Do we need to implement those client methods for flows (the same for `update_flow_configuration` and others), within our flow tools we can just use: 
```
await client.configuration_detail(
            component_id=ORCHESTRATOR_COMPONENT_ID,
            configuration_id=config_id
        )
  ```

vita-stejskal replied:
There are certainly pros and cons of this approach:

* pros -- more compact code in the flow tools; easier mocking in the tests
* cons -- clients cluttered with many function

Anyway, let's leave it as it is.

vita-stejskal commented on src/keboola_mcp_server/tools/flow.py:
```diff
@@ -0,0 +1,357 @@
+    flow_ids: Annotated[
+        Sequence[str],
+        Field(default_factory=tuple, description='The configuration IDs of the flows to retrieve.')
+    ] = tuple(),
```
The python default is needed when calling the function `retrieve_flows(...)` directly from python, e.g. in tests [here](https://github.com/keboola/mcp-server/blob/KAB-1057-flow-get-update-tool/tests/tools/test_flows.py#L423).

I am not sure if `default_factory` is really needed here, but I think that it affects the tool's "input JSON schema" and perhaps also the way how the tool's function is called via the MCP protocol. I admit that my understanding is limited in this area.
